crates/sui-core/src/execution_scheduler/balance_withdraw_scheduler/object_balance/naive_scheduler.rs

```diff
@@ -0,0 +1,211 @@
+    /// Tracks for each accumulator version, the object accounts that have unsettled withdraws.
+    tracked_accounts: BTreeMap<SequenceNumber, BTreeSet<AccumulatorObjId>>,
+    /// Tracks for each object account that has unsettled withdraws, the total amount of withdraws
+    /// that have not been settled yet, and the amount of withdraws that have not been settled
+    /// for each accumulator version.
+    unsettled_withdraws: BTreeMap<AccumulatorObjId, UnsettledWithdraws>,
+}
+
+#[derive(Default)]
+struct UnsettledWithdraws {
+    total_amount: u128,
+    unsettled_withdraw_map: BTreeMap<SequenceNumber, u128>,
+}
+
+pub(crate) enum ObjectBalanceWithdrawStatus {
+    SufficientBalance,
+    Pending(oneshot::Receiver<BalanceWithdrawStatus>),
+}
+
+impl ObjectBalanceWithdrawScheduler {
+    pub fn new(
+        balance_read: Arc<dyn AccountBalanceRead>,
+        starting_accumulator_version: SequenceNumber,
+    ) -> Self {
+        let (accumulator_version_sender, accumulator_version_receiver) =
+            watch::channel(starting_accumulator_version);
+        Self {
+            balance_read,
+            inner_state: Arc::new(Mutex::new(InnerState {
+                tracked_accounts: BTreeMap::new(),
+                unsettled_withdraws: BTreeMap::new(),
+            })),
+            accumulator_version_sender: Arc::new(accumulator_version_sender),
+            accumulator_version_receiver: Arc::new(accumulator_version_receiver),
+        }
+    }
+
+    pub fn schedule(
+        &self,
+        object_withdraws: BTreeMap<AccumulatorObjId, u64>,
+        accumulator_version: SequenceNumber,
+    ) -> ObjectBalanceWithdrawStatus {
+        let last_settled_version = *self.accumulator_version_receiver.borrow();
+        // This function is called during execution, which means this transaction is not committed yet,
+        // so the settlement transaction at the end of the same consensus commit cannot have settled yet.
+        assert!(accumulator_version >= last_settled_version);
+        // TODO: Should we use last_settled_version or accumulator_version here?
+        if self.check_balance_sufficient(&object_withdraws, accumulator_version) {
+            self.record_withdraws(&object_withdraws, accumulator_version);
+            return ObjectBalanceWithdrawStatus::SufficientBalance;
+        }
+        if accumulator_version == last_settled_version {
+            return Self::return_insufficient_balance();
+        }
+
+        // Spawn a task to wait for the last settled version to become accumulator_version,
+        // before we could check again.
+        let scheduler = self.clone();
+        let accumulator_version_sender = self.accumulator_version_sender.clone();
+        let (sender, receiver) = oneshot::channel();
+        tokio::spawn(async move {
+            let mut version_receiver = accumulator_version_sender.subscribe();
+            version_receiver
+                .wait_for(|v| *v == accumulator_version)
```

Comment:
according to the tokio docs, its possible for some values that are sent to the watch receiver to be skipped.
So this could wait forever. We should check for `*v >= accumulator_version` instead.